File: 0309-best-time-to-buy-and-sell-stock-with-cooldown/0309-best-time-to-buy-and-sell-stock-with-cooldown.java
class Solution {
    public int maxProfit(int[] prices) {
        if(prices.length == 1) return 0; 
        int[] dp = new int[prices.length]; 
        dp[0] = 0; 
        dp[1] = Math.max(prices[1] - prices[0], 0); 
        int max_diff = Math.max(-prices[0],-prices[1]); 
        
        for(int i = 2; i < prices.length; i++){
            dp[i] = Math.max(dp[i-1], prices[i] + max_diff); 
            max_diff = Math.max(max_diff, dp[i-2] - prices[i]); 
        }
        
        
        return dp[prices.length-1]; 
    }
}

File: 0309-best-time-to-buy-and-sell-stock-with-cooldown/0309-best-time-to-buy-and-sell-stock-with-cooldown.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size(); 
        if(n <= 1) return 0; 

        vector<vector<int>> dp(n,vector<int>(3,0)); 

        dp[0][0] = -prices[0]; 
        dp[0][1] = 0; 
        dp[0][2] = 0; 

        for(int i = 1; i < n; i++){
            dp[i][0] = max(dp[i-1][0], dp[i-1][1] - prices[i]);
            dp[i][1] = max(dp[i-1][1], dp[i-1][2]); 
            dp[i][2] = dp[i-1][0] + prices[i]; 
        }

        return max(dp[n-1][1],dp[n-1][2]); 
    }
};